sql/src/main/java/io/crate/execution/dsl/projection/WindowAggProjection.java

```diff
@@ -22,64 +22,79 @@
 
 import java.io.IOException;
 import java.util.ArrayList;
-import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
+import java.util.stream.Collectors;
 
 public class WindowAggProjection extends Projection {
 
     private final WindowDefinition windowDefinition;
     private final List<Symbol> standaloneWithInputs;
-    private final LinkedHashMap<WindowFunction, List<Symbol>> functionsWithInputs;
     private final ArrayList<Symbol> outputs;
+    private final ArrayList<WindowFunctionContext> windowFunctionContexts;
+    private final Set<WindowFunction> windowFunctions;
 
     public WindowAggProjection(WindowDefinition windowDefinition,
-                               LinkedHashMap<WindowFunction, List<Symbol>> functionsWithInputs,
+                               ArrayList<WindowFunctionContext> windowFunctionContexts,
                                List<Symbol> standaloneWithInputs) {
-        Set<WindowFunction> windowFunctions = functionsWithInputs.keySet();
+        this.windowFunctions = windowFunctionContexts.stream()
+            .map(WindowFunctionContext::function)
+            .collect(Collectors.toSet());
         assert windowFunctions.stream().noneMatch(Symbols.IS_COLUMN)
             : "Cannot operate on Reference or Field: " + windowFunctions;
         assert standaloneWithInputs.stream().noneMatch(Symbols.IS_COLUMN)
             : "Cannot operate on Reference or Field: " + standaloneWithInputs;
         this.windowDefinition = windowDefinition;
-        this.functionsWithInputs = functionsWithInputs;
+        this.windowFunctionContexts = windowFunctionContexts;
         this.standaloneWithInputs = standaloneWithInputs;
         outputs = new ArrayList<>(windowFunctions);
         outputs.addAll(standaloneWithInputs);
     }
 
-    public WindowAggProjection(StreamInput in) throws IOException {
+    WindowAggProjection(StreamInput in) throws IOException {
         windowDefinition = new WindowDefinition(in);
         standaloneWithInputs = Symbols.listFromStream(in);
         int functionsCount = in.readVInt();
-        functionsWithInputs = new LinkedHashMap<>(functionsCount, 1f);
+        windowFunctionContexts = new ArrayList<>();
         for (int i = 0; i < functionsCount; i++) {
             WindowFunction function = (WindowFunction) Symbols.fromStream(in);
+            Symbol filter;
+            if (in.getVersion().onOrAfter(Version.V_4_1_0)) {
+                filter = Symbols.fromStream(in);
+            } else {
+                filter = Literal.BOOLEAN_TRUE;
+            }
             List<Symbol> inputs = Symbols.listFromStream(in);
-            functionsWithInputs.put(function, inputs);
+            windowFunctionContexts.add(
+                new WindowFunctionContext(function, inputs, filter));
         }
-        outputs = new ArrayList<>(functionsWithInputs.keySet());
+        windowFunctions = windowFunctionContexts.stream()
```

Comment:
Could build the `windowFunctions` while going through the loop to avoid the `.stream().map().collectors()` overhead